platform/mbed_error_report.c

```diff
@@ -0,0 +1,215 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2013 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <stdlib.h>
+#include <stdarg.h>
+#include "rtx_os.h"
+#include "mbed_rtx.h"
+#include "hal/serial_api.h"
+#include "platform/mbed_error.h"
+#include "platform/mbed_error_report.h"
+
+#if DEVICE_SERIAL
+extern int stdio_uart_inited;
+extern serial_t stdio_uart;
+#endif
+
+
+/* Converts a uint32 to hex char string */
+static void value_to_hex_str(uint32_t value, char *hex_str)
+{
+    char hex_char_map[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+    int i = 0;
+    
+    //Return without converting if hex_str is not provided
+    if(hex_str == NULL) return;
+        
+    for(i=7; i>=0; i--) {
+        hex_str[i] = hex_char_map[(value & (0xf << (i * 4))) >> (i * 4)];
+    }
+}
+
+/* Converts a uint32 to dec char string */
+static void value_to_dec_str(uint32_t value, char *dec_str)
+{
+    char dec_char_map[] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9'};
+    int i = 0;
+    
+    //Return without converting if hex_str is not provided
+    if(dec_str == NULL) return;
+        
+    while(value > 0) {
+        dec_str[i++] = dec_char_map[value % 10];
+        value = value / 10;
+    }
+}
+
+/* Limited print functionality which prints the string out to 
+stdout/uart without using stdlib by directly calling serial-api 
+and also uses less resources 
+The fmtstr contains the format string for printing and for every %
+found in that it fetches a uint32 value from values buffer
+and prints it in hex format.
+*/
+void mbed_error_print(char *fmtstr, uint32_t *values)
+{
+#if DEVICE_SERIAL
```

Comment:
What about ITM and so on? I think we have nice abstraction there, so we can do it once and have it for all backends.

Reply:
I thought about that, but I found that we only have 2 devices(NRFs) with ITM support as of now, I think we need broader support so that it justifies adding more logic here to route it based on whats present and whats active.

Reply:
That's the case because we only needed it on these 2, due to small number of UARTs. I can envision a production device with multiple UARTs that are being used by WiFi, GPS, whatever and not having enough for debug.